### PR TITLE
Distributed claims

### PR DESCRIPTION
**Pull request by rohe**
Fixed bug in handling of distributed claims.

Thanks for contributing to this library! Please include the following check
list in your pull request submission (you can delete this message). If your
changes don't need a change log or documentation update, please ignore this.

- [ ] Any changes relevant to users are recorded in the `CHANGELOG.md`.
- [ ] The documentation has been updated, if necessary.

---

### CODE REVIEW

File: src/oic/oic/__init__.py [hunk 1]
```diff
@@ -1064,12 +1066,16 @@ def fetch_distributed_claims(self, userinfo, callback=None):
...
                 if "access_token" in spec:
                     _uinfo = self.do_user_info_request(
-                        token=spec["access_token"],
+                        method='GET', token=spec["access_token"],
                         userinfo_endpoint=spec["endpoint"])
                 else:
-                    _uinfo = self.do_user_info_request(
-                        token=callback(csrc),
-                        userinfo_endpoint=spec["endpoint"])
+                    if callback:
```
Comment (tpazderka) [inline src/oic/oic/__init__.py]: This part is not covered by tests.

File: src/oic/oic/__init__.py [hunk 2]
```diff
@@ -808,17 +808,19 @@ def user_info_request(self, method="GET", state="", scope="", **kwargs):
...
+                if token:
+                    try:
+                        _ttype = token.token_type
+                    except AttributeError:
+                        raise MissingParameter("Unspecified token type")
 
             if 'as_query_parameter' == _behav:
                 method = 'GET'
-            else:
+            elif token:
```
Comment (tpazderka) [inline src/oic/oic/__init__.py]: * Comparison to `None` should be explicit.
* In tests, this condition is always true, so a test with token = `None` should be added.

File: src/oic/oic/__init__.py (hunk 1 above)
Comment (rohe) [reply src/oic/oic/__init__.py]: Does the recent changes meet your requirements ?

Comment (tpazderka) [reply src/oic/oic/__init__.py]: Sorry for not responding earlier. I am having a busy week at work. I will have a look today.

File: src/oic/oic/__init__.py (hunk 2 above)
Comment (tpazderka) [reply src/oic/oic/__init__.py]: The comparison is still not explicit. It believe that it should be `elif token is not None:` as token is previously assigned `None`.

File: src/oic/oic/__init__.py [hunk 3]
```diff
@@ -808,17 +808,19 @@ def user_info_request(self, method="GET", state="", scope="", **kwargs):
...
             _token = uir["access_token"]
+            _ttype = ''
             try:
                 _ttype = kwargs["token_type"]
             except KeyError:
-                try:
-                    _ttype = token.token_type
-                except AttributeError:
-                    raise MissingParameter("Unspecified token type")
+                if token:
```
Comment (tpazderka) [inline src/oic/oic/__init__.py]: The same goes here: `if token is not None:`

File: src/oic/oic/__init__.py (hunk 1 above)
Comment (tpazderka) [reply src/oic/oic/__init__.py]: Here again, comparison to `None` and test without callback is missing.

File: src/oic/oic/__init__.py [hunk 4]
```diff
@@ -899,12 +901,13 @@ def do_user_info_request(self, method="POST", state="", scope="openid",
...
-        # Verify userinfo sub claim against what's returned in the ID Token
-        idt = self.grant[state].get_id_token()
-        if idt:
-            if idt['sub'] != res['sub']:
-                raise SubMismatch(
-                    'Sub identifier not the same in userinfo and Id Token')
+        if state:
+            # Verify userinfo sub claim against what's returned in the ID Token
+            idt = self.grant[state].get_id_token()
+            if idt:
```
Comment (tpazderka) [inline src/oic/oic/__init__.py]: This seems to never be triggered in tests.

**Comment by tpazderka** [comment]
I know this is a lot of tests to be added, but if you are already poking into the code, we might as well add test to cover the changes...

File: src/oic/oic/__init__.py (hunk 2 above)
Comment (rohe) [reply src/oic/oic/__init__.py]: I don't think that is good enough. token='' should be handled the same way as token=None
hence the use of ´<elif token>´ which implicitly include <token is not ''>.
So I'd rather not use not None since then I would have to have do
<elif token is not None and token is not ''>

Comment (tpazderka) [reply src/oic/oic/__init__.py]: And is there any case where token would be an empty string?
If yes, then OK, leave the code as is.

File: src/oic/oic/__init__.py (hunk 1 above)
Comment (rohe) [reply src/oic/oic/__init__.py]: OK now ? Need to get this out the door soon.

**Comment by tpazderka** [comment]
I would like to see more tests, but I understand that any progress is better than none.

**Comment by lwm** [comment]
> I would like to see more tests, but I understand that any progress is better than none.

Agreed!

If we feel like we're missing coverage, just create an issue with a link to this PR and we can track that. If a bug is ever raised, we will be motivated to come back to cover the remaining cases.

Nice work @rohe.

**Comment by rohe** [comment]
Sounds good to me.
So, can one of you merge the PR.
Since I'm the author I shouldn't do it myself.
After the merge I'd like us to release a new version to pypi. 
The next release of the OpenID Connect tool has a dependency on this PR and I want to get that out the door as soon as possible.

**Comment by lwm** [comment]
🍰 

> Since I'm the author I shouldn't do it myself.

Hey, you may be VIP but you're not getting out of merging your own PRs ;)

I won't get time to make a release, please someone, go for it and keep https://github.com/OpenIDC/pyoidc/issues/419 in mind.

**Comment by rohe** [comment]
Excellent new! Thanks !
VIP, lol :-)
I'll probably not have time to do a release today. At the earliest tomorrow afternoon.
